utils/ipl-sipwitch.1
.\" ipl-sipwtch - generate initial mysql database for sipwitch servers
.\" Copyright (C) 2018 Tycho Softworks
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH ipl-sipwitch "1" "January 2018" "SipWitchQt" "Tycho Softworks"
.SH NAME
ipl-sipwitch \- generate initial mysql database for sipwitch servers
.SH SYNOPSIS
.B ipl-sipwitch
.br
.SH DESCRIPTION
This utility is used to create and initially populate a mysql database with
initial authorization records when deploying a new sipwitch instance.  The
values needed to create a new database, including initial passwords, are
prompted for on the command line. The connection information for the mysql
server is typically read from the servers /etc/sipwitchqt.conf file.  Since the
utilities can be installed stand-alone on a separate machine, you can instead 
create a local config file with database connection info using the -c path option.

On Redhat systems you will have to use ``sudo gem install ruby-mysql2`` because
redhat does not seem to package ruby mysql2.  On debian you can use ``sudo apt-get install ruby-mysql2``.
.SH AUTHOR
.B ipl-sipwitch
was written by David Sugar <tychosoft@gmail.com>.
.SH "REPORTING BUGS"
Report bugs to tychosoft@gmail.com.
.SH COPYRIGHT
Copyright \(co 2018 Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



CHANGELOG
v0.3.3
* General code cleanup
* Build support for Hurd and OpenBSD added
* Fixed server hostname detection
* Improved schema support
* Ipl now works with ruby mysql2

v0.3.2
* No longer external vendoring openssl
* Redhat and fedora require updated copr repos
* Windows requires vcpkg and CMAKE_TOOLCHAIN_FILE to be set

v0.3.1
* Fix for freebsd build
* Improved openssl vendoring
* Packaging changes
* Layout of basic crypto support
* Restructured client sip code

v0.3.0
* Improved unit file
* Added security model checks; no extension keys implemented yet
* Admin and suspend states and modifications supported
* Added admin drop/disconnect
* Various ui improvements

v0.2.7
* Improved packaging support
* Updated copyright notices
* Utils use stand-alone to init mysql servers
* De-emphasizing sqlite beyond testing builds
* Device Management and verification support added 

v0.2.6
* Simplified expiration handling
* Re-arranged font selection
* Introduced desktop & audio notify setting
* Integrated login auto start behavior
* Speed dial ui support
* Call coverage setting support
* Call forwarding setting support
* Delay ringing option supported
* Per user and global auto answer settings
* Start of integrated voice mail
* Device keys introduced

v0.2.5
* Customization at build time
* Topic management in sessions
* Policy setting for main public room

v0.2.4
* Session/state based toolbar management
* Online status of all extensions shown
* Membership management for groups
* Server system notification messaging
* Delete authorizations supported

v0.2.3
* Schema changes
* Message expiration

v0.2.2
* Change password support
* Simple markdown support
* Zeroconfg support
* Message sync for existing extensions
* Functional testing support
* Improved adduser ui
* Expiration marking on client

v0.2.1
* Fix for crash sending offline messages
* Visual styling improvements and optimizations

V0.2.0
* IPL loading of tables, such as for external mysql db
* User classification of priviledges
* Ability to edit extension names from profile
* Clickable and highlighted urls in messages
* Offline message sync
* Cleaner db handling for login/logout and multiple dbs
* Client import/export
* Font picker for message view
* Admin can create users from client
* Phonebook expanded
* New message indicators

v0.1.4
* I want to send a message to someone
* I Want to receive messages
* I want to search my messages
* I want to logout, and back in
* I want to know what extensions I can login as
* I want the ui to become available in language of my choice
* I want to know what devices are on my extension
* I want to autostart when I login on linux
* sipwitchqt message laying and rendering engine introduced
* email and public uri properties added to contacts

v0.1.3
* Roster processing on client and server
* Search processing on sessions screen
* click feedback, doubleclick contacts
* vibrant and light icon schemes

v0.1.2
* Start of sessions page and status

v0.1.1
* udp for eXosip2 messaging support
* automatic authentication integrated
* Start of core messaging code
* Replacing json with xdp
* Start of contacts (phonebook) page

v0.1.0
* Core authentication complete 
* Core registration with expiration
* Extended registration support for sipwitchqt clients

v0.0.2
* More complete test harness debug config
* Reset and dialogs for client
* Release packaging for primary platforms

v0.0.1
* Initial development testing release



utils/ipl-sipwitch.rb
#!/usr/bin/env ruby
# Copyright (C) 2017-2018 Tycho Softworks
#
# This file is free software; as a special exception the author gives
# unlimited permission to copy and/or distribute it, with or without
# modifications, as long as this notice is preserved.

['optparse', 'io/console', 'digest', 'fileutils', 'mysql2'].each {|mod| require mod}

RESERVED_NAMES = ['operators', 'system', 'anonymous', 'nobody']
database = 'mysql'
digest_type = 'MD5'
first = '100'
last = '699'
config = '/nonexist'

Dir.chdir(File.dirname($0))

if STDIN.respond_to?(:noecho)
  def get_pass(prompt="Password: ")
    print prompt
    input = STDIN.noecho(&:gets).chomp
    exit unless input.size > 0
    print "\n"
    input
  end
else
  def get_pass(prompt="Password: ")
    input = `read -s -p "#{prompt}" password; echo $password`.chomp
    exit unless input.size > 0
    print "\n"
    input
  end
end

def get_input(*args)
  print(*args)
  input = gets.chomp
  exit unless input.size > 0
  input
end

def get_secret(type, user, domain, pass)
  case type
  when 'SHA-256'
    Digest::SHA256.hexdigest "#{user}:#{domain}:#{pass}"
  when 'SHA-512'
    Digest::SHA512.hexdigest "#{user}:#{domain}:#{pass}"
  else
    Digest::MD5.hexdigest "#{user}:#{domain}:#{pass}"
  end
end

OptionParser.new do |opts|
  opts.banner = 'Usage: ipl-sipwitch'

  opts.on('-2', '--sha256', 'use sha-256') do
    digest_type = 'SHA-256'
  end

  opts.on('-5', '--sha512', 'use sha-512') do
    digest_type = 'SHA-512'
  end

  opts.on('-c', '--config [PATH]', 'set config file') do |path|
    config = path
  end

  opts.on('-e', '--echo', 'echo commands to stdout') do
    echo_flag = true
  end
end.parse!
abort(opts.banner) if(ARGV.size > 0)

if !File.exists?(config) 
  config = '/etc/sipwitchqt.conf'
  config = '../testdata/service.conf' if File.exists?('../testdata/service.conf')
  config = '../userdata/service.conf' if File.exists?('../userdata/service.conf')
  localdb = '/var/lib/sipwitchqt/local.db'
end

abort("*** ipl-sipwitch: no config") unless File.exists?(config)
print "config file #{config} used\n"

section = nil
realm = nil
dbcfg = {:host => '127.0.0.1', :name => 'sipwitch', :port => '3306', :username => 'sipwitch', :password => nil}

File.open(config, 'r') do |infile; line, key, value|
  while(line = infile.gets)
    line.gsub!(/(^|\s)[#].*$/, '')
    case line.strip!
    when /^\[.*\]$/
      section = line[1..-2].downcase
    when /[=]/
      key, value = line.split(/\=/).each {|s| s.strip!}
      key.downcase!
      case section
      when nil
        case key
        when 'realm'
          realm = value
        when 'database'
          case value.downcase
          when /mysql/
            database = 'mysql'
          when /sqlite/
            database = 'sqlite'
          else
            abort("*** ipl-sipwitch: #{value}: unknown driver")
          end
        end
      when 'database'
        dbcfg[key.to_sym] = value
      end
    end
  end
end

# find schema
schema = "../Database/#{database}.sql"
schema = "../share/schemas/sipwitch-#{database}.sql" if File.exists?("/usr/share/schemas/sipwitch-#{database}.sql")
abort("*** ipl-sipwitch: no schema") unless File.exists?(schema)

# gather facts for initial config....

begin
  print "Creating realm \"#{realm}\" for #{database}\n" if (realm != nil && realm.size > 0)
  realm   = get_input "Server realm: " if realm === nil
  user    = get_input "Authorizing user: " 
  abort("*** ipl-sipwitch: #{user}: reserved name") if RESERVED_NAMES.include?(user)
  extnbr  = get_input "User extension:   "
  abort("*** ipl-sipwitch: ext must be #{first}-#{last}") unless extnbr >= first and extnbr <= last
  display = get_input "Display name:     "
  email = get_input   "Email address:    "
rescue Interrupt
  abort("")
  exit
end

begin
  print "\nEnter a password for #{extnbr} to authorize with\n"
  pass1    = get_pass "Password: "
  pass2    = get_pass "Verify:   "
  abort("*** ipl-sipwitch: passwords do not match") unless pass1 == pass2
  extpass = get_secret(digest_type,user,realm,pass1)

rescue Interrupt
  abort("^C")
  exit
end

IPL_COMMANDS = [
  "INSERT INTO Config(id, realm) VALUES(1,'#{realm}');",
  "INSERT INTO Authorize(authname,authtype,authaccess) VALUES('system', 'SYSTEM', 'LOCAL');",
  "INSERT INTO Authorize(authname,authtype,authaccess) VALUES('nobody', 'SYSTEM', 'LOCAL');",
  "INSERT INTO Authorize(authname,authtype,authaccess) VALUES('anonymous', 'SYSTEM', 'DISABLED');",
  "INSERT INTO Authorize(authname,authtype,authaccess) VALUES('operators', 'SYSTEM', 'PILOT');",
  "INSERT INTO Extensions(extnbr,authname,display) VALUES(0, 'operators', 'Operators');",
  "INSERT INTO Authorize(authname,authdigest,realm,secret,fullname,authtype,authaccess,email) VALUES('#{user}','#{digest_type}','#{realm}','#{extpass}','#{display}','USER','REMOTE','#{email}');",
  "INSERT INTO Extensions(extnbr,authname) VALUES(#{extnbr},'#{user}');",
  "INSERT INTO Endpoints(extnbr) VALUES(#{extnbr});",
  "INSERT INTO Admin(authname,extnbr) VALUES('system',#{extnbr});",
]

create = []
line = ""
dbcfg[:database] = dbcfg[:name]

begin
  db = Mysql2::Client.new(dbcfg)
  File.open(schema, 'r') do |infile; cmd|
    while(cmd = infile.gets)
      cmd.gsub!(/(^|\s)[-][-].*$/, '')
      case cmd.strip!
      when /^$/
        line.strip!
        db.query(line) if line.length > 1
        line = ""
      else
        line = "#{line} #{cmd}"
      end
    end
  end
  IPL_COMMANDS.each do |cmd|
    db.query(cmd)
  end
rescue Mysql2::Error => e
  abort("*** ipl-sipwitch: mysql error=#{e.errno},#{e.error}")
end

